File: web/src/figures/fig_1_4.ts
import Plotly from 'plotly.js-dist-min';
import { slider } from '../common/ui';

const controls = document.getElementById('controls')!;
const plot = document.getElementById('plot')!;
const caption = document.getElementById('caption')!;
caption.textContent = 'FIGURE 1.4 Constructive and destructive interference: ψtot and |ψtot|².';

const phaseSlider = slider(controls, 'Δφ', 0, 360, 0, 1);
const shiftSlider = slider(controls, 'shift', -10, 10, 0, 1);
const ampSlider = slider(controls, 'A2/A1', 0, 2, 1, 0.1);

const compute = () => {
  const phase = (parseFloat(phaseSlider.value) * Math.PI) / 180;
  const shift = parseFloat(shiftSlider.value) / 10;
  const amp2 = parseFloat(ampSlider.value);

  const xs: number[] = [];
  const psi1: number[] = [];
  const psi2: number[] = [];
  const psit: number[] = [];
  const rho: number[] = [];

  for (let i = 0; i <= 400; i++) {
    const x = -10 + (20 * i) / 400;
    const w = 2 * Math.PI * 0.6;
    const p1 = Math.sin(w * x);
    const p2 = amp2 * Math.sin(w * (x - shift) + phase);
    const pt = p1 + p2;
    xs.push(x);
    psi1.push(p1);
    psi2.push(p2);
    psit.push(pt);
    rho.push(pt * pt);
  }

  Plotly.newPlot(plot, [
    { x: xs, y: psi1, name: 'ψ₁', mode: 'lines', line: { color: '#1f6f8b' } },
    { x: xs, y: psi2, name: 'ψ₂', mode: 'lines', line: { color: '#5a4ea3' } },
    { x: xs, y: psit, name: 'ψtot', mode: 'lines', line: { color: '#b86b0f' } },
    { x: xs, y: rho, name: '|ψtot|²', mode: 'lines', line: { color: '#2ecc71' } }
  ], {
    margin: { l: 50, r: 20, t: 20, b: 40 },
    xaxis: { title: 'x' },
    yaxis: { title: 'Amplitude / Density' },
    legend: { orientation: 'h' }
  }, { displayModeBar: false });
};

[phaseSlider, shiftSlider, ampSlider].forEach((el) => el.addEventListener('input', compute));
compute();

File: pluto/fig_1_4.jl
### A Pluto.jl notebook ###
# v0.19.0

using PlutoUI
using WGLMakie
include("common/hydrogenic.jl")
using .Hydrogenic

md"""# Figure 1.4
Interference of two waves: psi1 + psi2 and |psi_tot|^2."""

@bind phase Slider(0:1:360, default=0)
@bind shift Slider(-10:1:10, default=0)
@bind amp2 Slider(0:0.1:2, default=1)

xs = range(-10, 10; length=400)
omega = 2*pi*0.6
psi1 = [sin(omega*x) for x in xs]
psi2 = [amp2*sin(omega*(x - shift/10) + phase*pi/180) for x in xs]
psit = [a+b for (a,b) in zip(psi1, psi2)]
rho = [x^2 for x in psit]

fig = Figure(resolution=(800,400))
ax = Axis(fig[1,1], xlabel="x", ylabel="Amplitude / Density")
lines!(ax, xs, psi1, color=:steelblue, label="psi1")
lines!(ax, xs, psi2, color=:purple, label="psi2")
lines!(ax, xs, psit, color=:orange, label="psi_tot")
lines!(ax, xs, rho, color=:green, label="|psi_tot|^2")
axislegend(ax)
fig
